### PR TITLE
Run the test suite as a GitHub Action

### PR DESCRIPTION
**Pull request by EmperorArthur**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,6 +29,43 @@ jobs:
...
+
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.x"
+
+    - name: Install OpenSCAD
+      run: |
+        sudo apt-get update
+        sudo apt-get install openscad
```
[the remark this answers is not in the document]
Comment (Ruudjhuu) [reply .github/workflows/ci.yml:47]: I do agree it is very old. Is it possible to use a fixed "nightly" version? I hate the ci/cd breaking due to the nightly being updated and unstable. I also don't like an openscad binary in this repo.

File: tests/openscad_runner.py [hunk 2]
```diff
@@ -121,20 +127,30 @@ class OpenScadRunner:
...
 
     def __init__(self, file_path: Path):
-        self.openscad_binary_path = self.WINDOWS_DEFAULT_PATH
+        self.openscad_binary_path = self.find_openscad_binary()
         self.scad_file_path = file_path
         self.image_folder_base = Path('.')
         self.camera_arguments = None
         self.parameters = None
 
+    @classmethod
```
Comment (Ruudjhuu) [inline tests/openscad_runner.py:136]: I do not see the need for this to be a classmethod. lets make it static.

File: tests/openscad_runner.py [hunk 3]
```diff
@@ -121,20 +127,30 @@ class OpenScadRunner:
...
+    def find_openscad_binary(cls) -> Path:
+        if cls.WINDOWS_DEFAULT_PATH.exists():
+            return cls.WINDOWS_DEFAULT_PATH
+        if cls.LINUX_DEFAULT_PATH.exists():
+            return cls.LINUX_DEFAULT_PATH
```
Comment (Ruudjhuu) [inline tests/openscad_runner.py:141]:
```suggestion
        exe= None
        if sys.platform == "windows":
            exe= cls.WINDOWS_DEFAULT_PATH
        if sys.platform == "linux":
            exe= cls.LINUX_DEFAULT_PATH
            
        if not exe or not exe.exists():
            raise OSError("openscad not found")
```

**Review by Ruudjhuu — DISMISSED** [review]

File: tests/openscad_runner.py (hunk 3 above)
Comment (EmperorArthur) [reply tests/openscad_runner.py:141]: Unfortunately, on my machine `sys.platform` returns 'win32'.  Maybe you meant to use `platform.system()`, and capitalize the first letter?

**Comment by EmperorArthur** [comment]
@pimlie I suspect it's partly also a methodological approach.

* I chose Python's unittest because it's a simple test framework that's easy to install and run.  I'm not opposed to other test frameworks at all, but feel that trying to homebrew a testing framework just to avoid a dependency causes far too many problems.
* Do your tests run under Windows?  It is very common for git bash to be installed, but it's not required for someone to use git.
* You might notice the Python tests can also use scad parameter files for some tests.  The difference is instead of one file per test, the parameter being tested is modified.
* This is explicitly doing unit tests at the library level.  Hence all the pictures of hole negatives.
* Image comparison would be nice, but I wonder if we could use csg export.  No major feelings either way.

# Future Work
* Realistically, the hole tests should be changed to do a matrix setup instead of brute force copy/paste, 
* I use Python's `unittest` framework not because it's the best test framework but because it's part of Python, and that is the easiest cross platform tool I am familiar with.

File: tests/openscad_runner.py (hunk 2 above)
Comment (EmperorArthur) [reply tests/openscad_runner.py:136]: If I'm reading the documentation correctly, excluding the internal changes to function signature, the largest difference is static methods would have to access variables by doing `OpenScadRunner.WINDOWS_DEFAULT_PATH`.

Meaning, class methods are equivalent to C++ static methods, and static methods are equivalent to a namespaced free function.

File: tests/openscad_runner.py (hunk 3 above)
Comment (EmperorArthur) [reply tests/openscad_runner.py:141]: I was debating on what you did or what's currently in the code.  I can go with your method but here's my logic:

* As I alluded to in my last comment, `sys.platform` has some gotchas, and [Python's docs](https://docs.python.org/3/library/sys.html#sys.platform) point to the `platform` module as a better choice.
* Portability and edge cases. [ `platform.system()`](https://docs.python.org/3/library/platform.html#platform.system) could return say "Darwin", and would mean Mac users can't run the test suite even if they had OpenSCAD installed in the default path.
* The current approach avoids having to import the platform module.  It's small, but is noticeable.
* Fewer number of lines, and same or fewer syscalls.

File: tests/openscad_runner.py [hunk 4]
```diff
@@ -99,19 +102,22 @@ class CameraRotations:
...
     '''Helper to run the openscad binary'''
+    camera_arguments: CameraArguments
     scad_file_path: Path
     openscad_binary_path: Path
     image_folder_base: Path
     parameters: Optional[dict]
     '''If set, a temporary parameter file is created, and used with these variables'''
 
-    WINDOWS_DEFAULT_PATH = 'C:\\Program Files\\OpenSCAD\\openscad.exe'
+    LINUX_DEFAULT_PATH = Path('/bin/openscad')
```
[the remark this answers is not in the document]
Comment (EmperorArthur) [reply tests/openscad_runner.py:112]: Does it in cases where you're not running through a shell?

File: tests/openscad_runner.py (hunk 2 above)
Comment (Ruudjhuu) [reply tests/openscad_runner.py:136]: I thought one of the differences is that class methods can modify the internal class state. A classmethod can return a class instance and thus is often used for factory functions. Because this function is doing nothing with the internal state of the cls object and just reading the "static" default path variables (they are not fields and are not tied to the lifetime of an instance) I think the function should be static. 

That said, I am ok with it either way as this is not the main code of this repo.

File: tests/openscad_runner.py (hunk 3 above)
Comment (Ruudjhuu) [reply tests/openscad_runner.py:141]: It is the train of thought while reading the code which is a (non) issue for me. The code you wrote has a different outcome on the existence of a path. Linux devices will check if the windows path exists which is a weird check. 

The real story of this code is different behaviour when running on different platforms. So why not implement the code exactly as the story why this code should be here.

On the note of fewer sys calls, you perform 2 disk io operations on Linux and one on windows. Which is slower than a platform call. But Python is already slow so who cares. 

Regarding the platform module, your probably right. 

Regarding Darwin, it is good to actively think about if the code also needs to be compatible on Darwin as it also differs here and there. This usecase it is the same which could be a coincidence. 

This is a nit-pick from me. I'm fine eighter way, it clearly works.

File: tests/openscad_runner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ruudjhuu) [reply tests/openscad_runner.py:112]: I am a fan of the absolute path as it won't depend on the users environment. If something is by default on Linux it does not mean we need to depend on it when it is not necessary. Defaults van change, users can do weird shit. We already have a big dependency on openscad. If they change the install location, it fails everywhere and we can change it.

**Comment by ntindle** [comment]
Just checking into what’s blocking this. Imo the comments are addressed and I’m always in favor of running tests in CI (even in a decreased capacity). Time is cheaper spent in CI than debugging after a break

**Comment by EmperorArthur** [comment]
> Just checking into what’s blocking this.

Merging requires at least one approving reviewer.  Which has not happened.  There are also some merge conflicts I need to address, and I have left it in draft status for the moment.

Right now, I'm focusing on fixing some bugs in the code itself, and getting CI/CD working again.

**Comment by ntindle** [comment]
Gotcha, no pressure just was trying to see if there was a place to help if I had free time

**Review by Ruudjhuu — APPROVED** [review]
Approved so you can move on
